Use zenoh for inter-daemon communication

phil-opp opened the pull request:
- Use `zenoh` for inter-daemon messaging
- Report error context from coordinator to CLI
- Assign unique ID to each daemon
- Rework default daemon selection
- Subscribe to required outputs of remote nodes when spawning dataflow
- Don't check dataflow in coordinator
- Log when sending AllNodesReady message
- Add more logging in daemon
- Improve log output
- Fix node stopping when error occurs during spawn
- Fix: Record node result when spawn fails
- Trim additional newlines at end of `stderr`

phil-opp commented:
@haixuanTao The change in https://github.com/dora-rs/dora/commit/57fd132998dbba035759b14a0aa6040e6478866c seems to prevent the CI from running here. Why did you switch from `paths-ignore` to `paths: !`? The documentation at https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#example-including-and-excluding-paths says:

> If you define a path with the `!` character, you must also define at least one path without the ! character. If you only want to exclude paths, use `paths-ignore` instead.

phil-opp commented:
I updated the PR description with usage instructions and information on the zenoh topic names.

haixuanTao commented:
Do you mind writing a markdown file within the multiple daemon example to document how to use this functionality?

haixuanTao commented:
Ok, sorry, Do you mind sharing the documentation for this but with a remote IP address? So, that configuration parameters are clearly explicited?

phil-opp commented:
You mean like the `Usage` section in the PR description?

haixuanTao commented:
I know this is not directly linked to this PR, but can we self-assign a new local-listen-port if the one we want to use is already taken. I feel like this is a common enough error that it should be avoided.

phil-opp commented:
> Yes!
>
> Also, could it be possible to outlines things that might not work?
> 
> It seems to me that this will not work by default if a daemon is under a NAT and another one isn't, right?
> 
> There is probably a need to spawn some form of zenoh routers.
> 
> Is there any additional things people should be aware of?

I pushed 2e6584c3 to include the usage instructions for distributed deployments in the README of the `multiple-daemons` example. I updated them to clarify that a zenoh router is needed when there are multiple sub-nets. I mostly referenced the zenoh docs for details since I didn't have time to try this myself yet.



> I think best would be if we can have shell blocks with ```bash ... ``` so that people can just copy paste it within there terminal.

I added/converted two commands blocks to `bash` blocks. 

I also updated the PR desciption with my changes.

phil-opp commented:
> I know this is not directly linked to this PR, but can we self-assign a new local-listen-port if the one we want to use is already taken. I feel like this is a common enough error that it should be avoided.

I don't see any use case for spawning multiple daemons on the same machine. This is just done for testing this on the CI. Or do you see any use case where this would be required?

Also, the `init_from_node_id` function used for dynamic nodes seems to be still hardcoded to the default port, so there is no way to use a non-standard `--local-listen-port` at the moment anyway... We can of course make it configurable through an additional function argument, but again, I'm not sure if there is really a use cause for this...

haixuanTao commented:
So it seems that this change does not use the `pub-sub` of the communication layer in order to create an abstraction layer of the communication pattern, and deeply integrate zenoh within dora daemon library.

Are we ok making zenoh basically the only pubsub layer for remote nodes?

Also, isn't zenoh topic going to fail if the node id contains a slash in the likes of node_id: `workstation_1/robot_1`? And same thing for the output? 

Using slash within ros topic is fairly common so it seems that this might create some conflict later on.

phil-opp commented:
> So it seems that this change does not use the `pub-sub` of the communication layer in order to create an abstraction layer of the communication pattern, and deeply integrate zenoh within dora daemon library.
> 
> Are we ok making zenoh basically the only pubsub layer for remote nodes?

Yes, this change doesn't use our old pub-sub communication layer abstraction. I thought about it, but our old code doesn't support `async` operations and was also designed for older versions of zenoh. So it was much simpler to just use zenoh directly.

However, I wouldn't say that we "deeply integrate" zenoh this way. We can easily update the code to support additional pub-sub implementations in the future. Of course, we could already try to create some abstraction that _hopefully_ works with other pub-sub implementations, but I think this is much easier to do when we have a concrete alternative pub-sub system we want to support (e.g. MQTT). Then we clearly see the differences in functionality that we need to abstract over.

The only thing that ties us to zenoh is that we support complex distributed networks using the `ZENOH_CONFIG` environment variable. This of course only works as long as keep using zenoh in the daemon. We could rename this env variable to `_UNSTABLE_ZENOH_CONFIG` to signal that this is just an experimental feature that we're not commited to yet.

phil-opp commented:
I opened a PR to forbid `/` characters in node IDs: https://github.com/dora-rs/dora/pull/785

phil-opp commented:
I think zenoh scouting uses multicast messages, which might be filtered in some networks (e.g. a guest WiFi). What kind of network are you using?

haixuanTao commented:
So, actually after double checking it seems to not work specifically when using wired connection on MacOS

haixuanTao commented:
Ok so actually it works but outside of VSCode. Using it within VSCode does not work on MacOS

phil-opp commented:
That's interesting. Is there any special environment used in VSCode on macOS. I wouldn't expect that VSCode has an influence on the network environment, assuming that everything is run locally.

haixuanTao commented:
I think it's linked to MacOS privacy settings and probably disabling  VSCode from listening to incoming internet traffic. I'm going to double check

haixuanTao commented:
Also, just talked with zenoh and they seem to not support edge->cloud connection:
![image](https://github.com/user-attachments/assets/5257d3e0-0d9f-4de4-a325-914d48132297)

phil-opp commented:
Copying the solution from the discord thread:

> Ah ok so i figured it out. I need to pass the given tcp port from the zenoh session in the cloud:
>
> ```
>        if let Some(addr) = coordinator_addr {
>            zenoh_config
>                .insert_json5(
>                    "connect/endpoints",
>                    &format!(r#"["tcp/{}:33505"]"#, addr.ip()),
>                )
>                .unwrap();
>        }
> ```
